### DISCUSSION TITLE
Warn when a test doesn't assert anything

### DISCUSSION THREAD

**Discussion by djmattyg007** [pytest-dev/pytest#9004] [Q&A]
Is there a way to make pytest warn or even fail when a test doesn't assert anything? PHPUnit calls these "risky" or "useless" tests.

https://phpunit.readthedocs.io/en/9.5/risky-tests.html

**Comment by RonnyPfannschmidt** [comment]
AFAIK The recently added assertion Pass hook may help (im on mobile so I'll skip on adding the documentation link)

**Reply by djmattyg007** [reply]
Are you referring to this?

https://github.com/pytest-dev/pytest/pull/8967

**Reply by RonnyPfannschmidt** [reply]
Yup

**Reply by djmattyg007** [reply]
Would you be able to elaborate on how I can use that to check for tests that don't assert anything?

**Reply by RonnyPfannschmidt** [reply]
any passed test which does not invoke the hook, didnt have a tracked assertion
a requirement however is working assertion rewrite

its not integrated with unittest assertions

**Comment by graingert** [comment]
I think it would be better as a feature request in https://pypi.org/project/flake8-pytest-style/

however there's a number of different ways to assert things about a system in a test that are not supported by the assertion pass hook eg `some_mock_obj.assert_called_with` fwiw I prefer `assert some_mock_obj.mock_calls == [...]` which is supported.

@djmattyg007 I've not been impacted by this specific misconfiguration. Only when tests are simply not executed at all (due to being async/deferred) [and that I solve by enforcing 100% coverage on my test suite ](https://nedbatchelder.com/blog/201908/dont_omit_tests_from_coverage.html)

**Reply by djmattyg007** [reply]
I don't understand why you think it's a "misconfiguration". What a test does or doesn't do has nothing to do with how I've configured pytest.

I also don't understand how the assertion pass hook is supposed to help me. Just pointing me at an unreleased-but-merged PR doesn't magically help me reach your level of understanding.

**Reply by graingert** [reply]
> I also don't understand how the assertion pass hook is supposed to help me. Just pointing me at an unreleased-but-merged PR doesn't magically help me reach your level of understanding.

I think you may be responding to the wrong post, I'm suggesting that you don't use pytest_assertion_pass. I think what @RonnyPfannschmidt is suggesting you do is to check that if a test fails then you know it ran an assertion. If a test passes and pytest_assertion_pass was not run for that test then it made no assertions.

**Reply by djmattyg007** [reply]
By the sounds of it, the answer to my question is "no, there isn't a way".

Which is perfectly fine, I don't know why people think half-explaining roundabout hacks is a good way of answering a question.

**Reply by graingert** [reply]
> By the sounds of it, the answer to my question is "no, there isn't a way".

No that's not what I'm saying

> Which is perfectly fine, I don't know why people think half-explaining roundabout hacks is a good way of answering a question.

I'm not sure what you mean.

**Reply by djmattyg007** [reply]
In my eyes, me having to write a whole bunch of bespoke code does not constitute the framework supporting a feature. It means the framework does not support the feature.

**Reply by nicoddemus** [reply]
@djmattyg007 the framework does not support the feature.

**Reply by nicoddemus** [reply]
Note that the way you phrased the question:

> Is there a way to make pytest warn or even fail when a test doesn't assert anything?

Conveys that you were asking for advice, not a clear-cut "does pytest supports this out of the box?".
